src/data/polynomial/ring_division.lean

```diff
@@ -207,6 +228,61 @@ end no_zero_divisors
 section no_zero_divisors
 variables [comm_semiring R] [no_zero_divisors R] {p q : R[X]}
 
+lemma monic.irreducible_iff (hp : p.monic) :
+  irreducible p ↔ p ≠ 1 ∧ ∀ a b, p = a * b →
+    a.monic → b.monic → 0 < a.nat_degree → 0 < b.nat_degree →
+    a.nat_degree + b.nat_degree = p.nat_degree → is_unit a ∨ is_unit b :=
+begin
+  rw irreducible_iff,
+  refine and_congr (not_congr ⟨hp.eq_one_of_is_unit, λ h, h.symm ▸ is_unit_one⟩)
+    ⟨forall₃_imp $ λ _ _ _ h _ _ _ _ _, h, _⟩,
+  rintro h a b rfl, nontriviality R,
+  rw [monic.def, leading_coeff_mul] at hp,
+  let u := units.mk_of_mul_eq_one _ _ hp,
+  obtain hd|hda := @eq_zero_or_pos _ _ a.nat_degree, left, swap,
+  obtain hd|hdb := @eq_zero_or_pos _ _ b.nat_degree, rw mul_comm at hp, right, swap 3,
+  iterate 2
+  { rw [eq_C_of_nat_degree_eq_zero hd, ← hd], exact (is_unit_of_mul_eq_one _ _ hp).map C },
+  have he := _, have ha := _, have hb := _,
+  refine (h (C u.inv * a) (C u.val * b) he ha hb _ _ $ by rw [he, ha.nat_degree_mul hb]).imp _ _,
+  iterate 2 { exact is_unit_of_mul_is_unit_right },
+  { rw nat_degree_C_mul, exacts [hda, u⁻¹.ne_zero] },
+  { rw nat_degree_C_mul, exacts [hdb, u.ne_zero] },
+  iterate 2
+  { rw [monic.def, leading_coeff_mul, leading_coeff_C], apply_rules [u.inv_val, u.val_inv] },
+  simp_rw [mul_left_comm, ← mul_assoc, ← C_mul, u.val_inv, C_1, one_mul],
+end
+
+lemma monic.not_irreducible_iff_exists_add_mul_eq_coeff (hm : p.monic) (hnd : p.nat_degree = 2) :
+  ¬ irreducible p ↔ ∃ c₁ c₂, p.coeff 0 = c₁ * c₂ ∧ p.coeff 1 = c₁ + c₂ :=
+begin
+  refine ⟨λ hred, _, _⟩, swap,
+  { rw irreducible_iff, push_neg, nontriviality R,
+    rintro ⟨c₁, c₂, hmul, hadd⟩ -,
+    refine ⟨X + C c₁, X + C c₂, _, _⟩,
+    { rw [p.as_sum_range_C_mul_X_pow, hnd], simp_rw [finset.sum_range_succ],
+      rw [finset.sum_range_zero, hmul, hadd, C_mul, C_add], nth_rewrite 0 ← hnd,
+      rw [← leading_coeff, hm.leading_coeff, C_1], ring },
+    split; { refine mt degree_eq_zero_of_is_unit _, rw degree_X_add_C, exact one_ne_zero } },
+  nontriviality R,
+  rw hm.irreducible_iff at hred, push_neg at hred,
+  obtain ⟨a, b, rfl, hma, hmb, hda, hdb, hd, hua, hub⟩ := hred _, swap,
+  { apply_fun nat_degree, rw [hnd, nat_degree_one], rintro ⟨⟩ },
+  rw hnd at hd, have := (nat.le_add_right _ _).trans_eq hd,
+  interval_cases a.nat_degree with hda', swap,
+  { rw [hda', add_right_eq_self] at hd, exact (hdb.ne' hd).elim },
+  rw hda' at hd, replace hd := add_left_cancel hd,
+  use [a.coeff 0, b.coeff 0, by rw mul_coeff_zero],
+  convert hma.next_coeff_mul hmb,
+  all_goals { simp only [next_coeff, hnd, hda', hd], rw if_neg, dec_trivial },
+end
+
+lemma irreducible_of_monic (hp : p.monic) (hp1 : p ≠ 1) :
+  irreducible p ↔ ∀ f g : R[X], f.monic → g.monic → f * g = p → f = 1 ∨ g = 1 :=
+by refine ⟨λ h f g hf hg hp, (h.2 f g hp.symm).imp hf.eq_one_of_is_unit hg.eq_one_of_is_unit,
+  λ h, hp.irreducible_iff.2 ⟨hp1, λ f g he hf hg _ _ _, (h f g hf hg he.symm).imp _ _⟩⟩;
+  exact λ h, h.symm ▸ is_unit_one
```

Comment:
It took a while, but here's how I would write this part of the file. It's slightly longer in line count but has a bonus lemma `monic.nat_degree_eq_zero_iff`, and it doesn't need such a complicated lemma for irreducibility. (If moreover we restate `monic.irreducible_iff_nat_degree` to have `(hp1 : p ≠ 1)` as a hypothesis, the total line count will be exactly the same.)

```suggestion
lemma irreducible_of_monic (hp : p.monic) (hp1 : p ≠ 1) :
  irreducible p ↔ ∀ f g : R[X], f.monic → g.monic → f * g = p → f = 1 ∨ g = 1 :=
begin
  refine ⟨λ h f g hf hg hp, (h.2 f g hp.symm).elim (or.inl ∘ hf.eq_one_of_is_unit)
    (or.inr ∘ hg.eq_one_of_is_unit), λ h, ⟨hp1 ∘ hp.eq_one_of_is_unit, λ f g hfg,
      (h (g * C f.leading_coeff) (f * C g.leading_coeff) _ _ _).elim
        (or.inr ∘ is_unit_of_mul_eq_one g _) (or.inl ∘ is_unit_of_mul_eq_one f _)⟩⟩,
  { rwa [monic, leading_coeff_mul, leading_coeff_C, ←leading_coeff_mul, mul_comm, ←hfg, ←monic] },
  { rwa [monic, leading_coeff_mul, leading_coeff_C, ←leading_coeff_mul, ←hfg, ←monic] },
  { rw [mul_mul_mul_comm, ←C_mul, ←leading_coeff_mul, ←hfg, hp.leading_coeff, C_1, mul_one,
        mul_comm, ←hfg] },
end

lemma monic.nat_degree_eq_zero_iff (hm : monic p) : nat_degree p = 0 ↔ p = 1 :=
⟨λ h, by rw [eq_C_of_nat_degree_eq_zero h, ← h, hm.coeff_nat_degree, map_one],
 λ h, by rw [h, nat_degree_one]⟩

lemma monic.irreducible_iff_nat_degree (hp : p.monic) :
  irreducible p ↔ p ≠ 1 ∧ ∀ a b : R[X], a.monic → b.monic → a * b = p →
    nat_degree a = 0 ∨ nat_degree b = 0 :=
begin
  by_cases hp1 : p = 1,
  { simp [hp1] },
  rw [irreducible_of_monic hp hp1, and_iff_right hp1],
  split;
    rintros h a b ha hb hp;
    simpa only [ha.nat_degree_eq_zero_iff, hb.nat_degree_eq_zero_iff] using h a b ha hb hp
end

lemma monic.not_irreducible_iff_exists_add_mul_eq_coeff (hm : p.monic) (hnd : p.nat_degree = 2) :
  ¬ irreducible p ↔ ∃ c₁ c₂, p.coeff 0 = c₁ * c₂ ∧ p.coeff 1 = c₁ + c₂ :=
begin
  casesI subsingleton_or_nontrivial R,
  { simpa [polynomial.nat_degree_of_subsingleton] using hnd },
  rw hm.irreducible_iff_nat_degree,
  push_neg,
  rw imp_iff_right,
  split,
  { rintros ⟨a, b, ha, hb, rfl, ha0, hb0⟩,
    use [a.coeff 0, b.coeff 0, by rw mul_coeff_zero],
     rw [nat_degree_mul ha.ne_zero hb.ne_zero] at hnd,
    rw ← pos_iff_ne_zero at ha0 hb0,
    have hda : a.nat_degree = 1,
    { exact le_antisymm (le_of_add_le_add_right (hnd.le.trans (add_le_add_left hb0 _))) ha0 },
    have hdb : b.nat_degree = 1 := add_left_cancel (by rwa hda at hnd),
    convert ha.next_coeff_mul hb,
    all_goals { simp only [next_coeff, hnd, hda, hdb, hnd, nat_degree_mul ha.ne_zero hb.ne_zero,
                           if_neg two_ne_zero, if_neg one_ne_zero] } },
  { rintros ⟨c₁, c₂, hmul, hadd⟩,
    refine ⟨X + C c₁, X + C c₂, monic_X_add_C _, monic_X_add_C _, _, _, _⟩,
    { rw [p.as_sum_range_C_mul_X_pow, hnd, finset.sum_range_succ, finset.sum_range_succ,
          finset.sum_range_one, ← hnd, hm.coeff_nat_degree, hnd, hmul, hadd, C_mul, C_add, C_1],
      ring },
    all_goals { rw nat_degree_X_add_C, exact one_ne_zero } },
  { rintro rfl,
    simpa using hnd },
end
```

Reply:
@Vierkantor Your refactor looks great; I mostly adopted it (except I added a version of `irreducible_iff_nat_degree` using `finset.Ioc` which slightly simplifies the quadratic proof) and added you as an coauthor. Notice that `monic.nat_degree_eq_zero_iff` already exists as [polynomial.monic.nat_degree_eq_zero_iff_eq_one](https://leanprover-community.github.io/mathlib_docs/data/polynomial/monic.html#polynomial.monic.nat_degree_eq_zero_iff_eq_one).